core/consensus/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -0,0 +1,214 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "consensus/synchronizer/impl/synchronizer_impl.hpp"
+
+#include <boost/assert.hpp>
+#include <boost/optional.hpp>
+#include "blockchain/block_tree_error.hpp"
+#include "network/types/block_announce.hpp"
+#include "network/types/block_request.hpp"
+#include "network/types/block_response.hpp"
+
+OUTCOME_CPP_DEFINE_CATEGORY(kagome::consensus, SynchronizerImpl::Error, e) {
+  using E = kagome::consensus::SynchronizerImpl::Error;
+  switch (e) {
+    case E::INVALID_BLOCK:
+      return "peer has sent an invalid or incomplete block";
+    case E::PEER_RETURNED_NOTHING:
+      return "peer has not sent any blocks in response; either it does not "
+             "have a required hash or it is in non-finalized fork";
+    case E::NO_SUCH_PEER:
+      return "no peer with such ID found";
+  }
+  return "unknown error";
+}
+
+namespace kagome::consensus {
+  using network::BlockRequest;
+  using network::BlockResponse;
+
+  SynchronizerImpl::SynchronizerImpl(
+      std::shared_ptr<blockchain::BlockTree> block_tree,
+      std::shared_ptr<blockchain::BlockHeaderRepository> blocks_headers,
+      std::shared_ptr<network::NetworkState> network_state,
+      SynchronizerConfig config,
+      common::Logger log)
+      : block_tree_{std::move(block_tree)},
+        blocks_headers_{std::move(blocks_headers)},
+        network_state_{std::move(network_state)},
+        config_{config},
+        log_{std::move(log)} {
+    BOOST_ASSERT(block_tree_);
+    BOOST_ASSERT(blocks_headers_);
+    BOOST_ASSERT(network_state_);
+    BOOST_ASSERT(log_);
+  }
+
+  void SynchronizerImpl::start() {
+    network_state_->peer_server->onBlocksRequest(
+        [self{shared_from_this()}](
+            const BlockRequest &request) -> outcome::result<BlockResponse> {
+          return self->processRequest(request);
+        });
+  }
+
+  void SynchronizerImpl::announce(const primitives::BlockHeader &block_header) {
+    auto announce = network::BlockAnnounce{block_header};
+    for (const auto &peer_client : network_state_->peer_clients) {
+      peer_client.second->blockAnnounce(
+          announce, [self{shared_from_this()}, peer_client](auto &&res) {
+            if (!res) {
+              self->log_->error("cannot send a block announce to peer {}: {}",
+                                peer_client.first.toBase58(),
+                                res.error().message());
+            }
+          });
+    }
+  }
+
+  void SynchronizerImpl::requestBlocks(const libp2p::peer::PeerInfo &peer,
```

Comment:
do not expose libp2p :)

Reply:
anyway you retrieve peer id from it eventually